### ISSUE TITLE
docs(skore): Custom metric cache is based only on function name

### ISSUE THREAD

**Issue by divakaivan** [probabl-ai/skore#2060]
## Context 

If we run the below code for a custom metric, I get my expected output

```python
# cell 1
from sklearn.datasets import load_diabetes
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error
from skore import train_test_split
from skore import EstimatorReport
```

```python
# cell 2
X, y = load_diabetes(return_X_y=True)
split_data = train_test_split(X=X, y=y, random_state=0, as_dict=True)
regressor = Ridge()
report = EstimatorReport(regressor, **split_data)
```

```python
# cell 3
def metric_function(y_true, y_pred):
    return {"output": mean_absolute_error(y_true, y_pred)}

report.metrics.custom_metric(
    metric_function=metric_function,
    response_method="predict",
)
# {"output" 44.922683494675425}
```

If I change my mind as to how the output looks like (or even the calculation)

```python
# cell 3 (now change metric_function's calculation)
def metric_function(y_true, y_pred):
    return 0.2 # or just mean_absolute_error(y_true, y_pred)

report.metrics.custom_metric(
    metric_function=metric_function,
    response_method="predict",
)
```

I will still get

```
{"output" 44.922683494675425}
```

## Alternatives

Atm we cache the metric function based [on the function name](https://github.com/probabl-ai/skore/blob/7251e6974c9b81757c1433edf9e1dcb2f60ce502/skore/src/skore/_sklearn/_estimator/metrics_accessor.py#L451C9-L455C10)

```python
# from skore/src/skore/_sklearn/_estimator/metrics_accessor.py
cache_key_parts: list[Any] = [
    self._parent._hash,
    metric_fn.__name__,
    data_source,
]
```

And indeed if I change my function name

```python
def metric_function_CHANGED(y_true, y_pred):
    return 0.2 

report.metrics.custom_metric(
    metric_function=metric_function_CHANGED,
    response_method="predict",
)
# now I get 0.2
```

Would some hashing based on the function contents be better? For instance:

```python
import hashlib
import inspect

def foo1(x):
    return str(x)

def hash_function(fn) -> str:
    hasher = hashlib.sha256()
    hasher.update(fn.__code__.co_code)
    return hasher.hexdigest()

hash1 = hash_function(foo1)

# change function body to return something different but keep the same name
def foo1(x):
    return x

hash2 = hash_function(foo1)
hash1 == hash2 # False
```

Or use `id(metric_function)`

❗ a user on discord (pyphi) also mentioned they faced a similar issue when trying to use custom metrics. exact msg: *Do you know if we can disable cahing? I got into issues updating the metric function, before I realise skore is chaching stuff!*

**Comment by thomass-dev** [comment]
IMO, this is a tricky problem: hashing the custom metric function will not solve issues when some of its dependencies changed.

For instance, we will not be able to detect a change from

```
def function():
    return 1

def metric_function(y_true, y_pred):
    return function()
```

to

```
def function():
    return 2

def metric_function(y_true, y_pred):
    return function()
```

**Comment by thomass-dev** [comment]
> Or use id(metric_function)

IMO, correct me if i'm wrong, it's not a good idea: the id is dependent of your session. After a restart of your python program, the id of the same function is not guarantee to be the same.

So if you dump your report using default pickle options, the cache will be invalid on reloading.

**Comment by divakaivan** [comment]
> IMO, correct me if i'm wrong, it's not a good idea: the id is dependent of your session. After a restart of your python program, the id of the same function is not guarantee to be the same.

Maybe, I'm not doing the check correctly but seems the cache is valid? Here is the code I used

In `skore/src/skore/_sklearn/_estimator/metrics_accessor.py` I changed this

```python
        cache_key_parts: list[Any] = [
            self._parent._hash,
            id(metric_fn), # metric_fn.__name__, 
            data_source,
        ]
```

and then:

```python
### cell 1
from sklearn.datasets import load_diabetes
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error
from skore import train_test_split
from skore import EstimatorReport
import time

### cell 2
X, y = load_diabetes(return_X_y=True)
split_data = train_test_split(X=X, y=y, random_state=0, as_dict=True)
regressor = Ridge()
report = EstimatorReport(regressor, **split_data)

### cell 3
start = time.perf_counter()

report.metrics.r2()

end = time.perf_counter()
time_no_cache = end - start
print(f"{time_no_cache:.6f}") # 0.000663

### cell 4
runs = 10
total = 0

for _ in range(runs):
    start = time.perf_counter()
    report.metrics.r2()
    end = time.perf_counter()
    total += (end - start)

avg_time = total / runs
print(f"{avg_time:.6f} seconds") # 0.000061 seconds

### cell 5
import pickle

with open("report.pkl", "wb") as f:
    pickle.dump(report, f)

### restarted the kernel and ran the below 2 cells###

### cell 6
import pickle

with open('report.pkl', 'rb') as f:
    loaded_report = pickle.load(f)

### cell 7
import time 

runs = 10
total = 0

for _ in range(runs):
    start = time.perf_counter()
    loaded_report.metrics.r2()
    end = time.perf_counter()
    total += (end - start)

avg_time = total / runs
print(f"{avg_time:.6f} seconds") # 0.000082 seconds
```

**Comment by glemaitre** [comment]
I find it is a corner case:

- a user define a metric with a name
- create a report passing this function
- redefine the metric
- reuse the same exact report instance to compute the new metric

I think that it would rarely happen in this way. But it is true that the current caching mechanism is not robust to such case.

I think that we could look at the way joblib is caching computation but there is anyway caveats: https://joblib.readthedocs.io/en/latest/memory.html#gotchas

**Comment by glemaitre** [comment]
OK so I would advocate to at least document this behaviour. We have a specific example dedicated to the cache where we could show this "gotcha" pattern.

**Comment by cakedev0** [comment]
We call  `joblib.hash` on callables in `deep_key_sanitize` now. So I think this issue is resolved.

(feel free to re-open the issue if I'm wrong)

**Comment by auguste-probabl** [comment]
`joblib.hash` is not enough:
```python
>>> import joblib
>>> def f(x): 123
>>> joblib.hash(f)
'1dbf95b7d3cae635016b6b904637b863'
>>> def f(x): 456
>>> joblib.hash(f)
'1dbf95b7d3cae635016b6b904637b863'
```

**Comment by auguste-probabl** [comment]
That said, we discussed the possibility of supporting only "scorer" callables, i.e. of the form `(estimator, X, y) -> score`, and no longer supporting "metric" callables, i.e. `(y_true, y_pred) -> score`, because the former is more flexible and likely the most useful way of defining a custom metric. If we do this, then caching custom metrics will look a bit different because we won't be able to tell which prediction function is run inside of the scorer. There is an exception for `_BaseScorer` scorers (those constructed with `sklearn.metrics.make_scorer`) because those can be unpacked into a `(y_true, y_pred) -> score` metric, where we can reuse our cached predictions. In the meantime, documenting the current behaviour is probably the easiest way to close this issue.

Also, see https://github.com/probabl-ai/skore/pull/2725#issuecomment-4207167799

**Comment by auguste-probabl** [comment]
`custom_metric` no longer exists, and `report.metrics.add` now errors if a metric is added that has the same name as a registered one.
